packages/assets-controllers/src/MultichainBalancesController/MultichainBalancesController.ts

```diff
@@ -0,0 +1,396 @@
+    );
+  }
+
+  /**
+   * Starts the polling process.
+   */
+  start(): void {
+    this.#tracker.start();
+  }
+
+  /**
+   * Stops the polling process.
+   */
+  stop(): void {
+    this.#tracker.stop();
+  }
+
+  /**
+   * Tracks an account to get its balances.
+   *
+   * @param account - The account to track.
+   */
+  #trackAccount(account: InternalAccount): void {
+    if (!this.#isNonEvmAccount(account)) {
+      // Nothing to do here for EVM accounts
+      return;
+    }
+
+    const updateTime =
+      // @ts-expect-error - For the moment we are only tracking non-EVM accounts and this is
+      // checked with the method `#isNonEvmAccount`. We can ignore this error since
+      // eip155:eoa and eip155:erc4337 account type balances are not tracked here.
+      BALANCE_UPDATE_INTERVALS[this.#getAccount(account.id).type];
+    this.#tracker.track(account.id, updateTime);
+  }
+
+  /**
+   * Lists the multichain accounts coming from the `AccountsController`.
+   *
+   * @returns A list of multichain accounts.
+   */
+  #listMultichainAccounts(): InternalAccount[] {
+    return this.messagingSystem.call(
+      'AccountsController:listMultichainAccounts',
+    );
+  }
+
+  /**
+   * Lists the accounts that we should get balances for.
+   *
+   * Currently, we only get balances for non-EVM accounts.
+   *
+   * @returns A list of accounts that we should get balances for.
+   */
+  #listAccounts(): InternalAccount[] {
+    const accounts = this.#listMultichainAccounts();
+
+    return accounts.filter((account) => this.#isNonEvmAccount(account));
+  }
+
+  /**
+   * Get a non-EVM account from its ID.
+   *
+   * @param accountId - The account ID.
+   * @returns The non-EVM account.
+   */
+  #getAccount(accountId: string): InternalAccount {
+    const account: InternalAccount | undefined =
+      this.#listMultichainAccounts().find(
+        (multichainAccount) => multichainAccount.id === accountId,
+      );
+
+    if (!account) {
+      throw new Error(`Unknown account: ${accountId}`);
+    }
+    if (!this.#isNonEvmAccount(account)) {
+      throw new Error(`Account is not a non-EVM account: ${accountId}`);
+    }
+    return account;
+  }
+
+  /**
+   * Updates the balances of one account. This method doesn't return
+   * anything, but it updates the state of the controller.
+   *
+   * @param accountId - The account ID.
+   */
+  async #updateBalance(accountId: string) {
+    const account = this.#getAccount(accountId);
+    const partialState: MultichainBalancesControllerState = { balances: {} };
+
+    const scope = getScopeForAddress(account);
+    const assetsList = this.#networkAssetsMap[scope];
+
+    if (account.metadata.snap) {
+      partialState.balances[account.id] = await this.#getBalances(
+        account.id,
+        account.metadata.snap.id,
+        assetsList,
+      );
+    }
+
+    this.update((state: Draft<MultichainBalancesControllerState>) => ({
+      ...state,
```

Comment:
Is there a reason that we need to spread here? If I recall correctly, when using an immer Draft you can directly mutate the state object.

Reply:
Good point [c3f273d](https://github.com/MetaMask/core/pull/4965/commits/c3f273dba82851c1300ce1ed7e116b2c312a00c1)